File: backend/app/schemas/video.py
from pydantic import BaseModel, ConfigDict, Field


class VideoMetadata(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    filename: str
    duration: float = Field(ge=0)
    width: int = Field(gt=0)
    height: int = Field(gt=0)
    fps: float = Field(gt=0)
    video_codec: str = Field(alias="videoCodec")
    audio_codec: str | None = Field(default=None, alias="audioCodec")
    has_audio: bool = Field(alias="hasAudio")
    size_bytes: int = Field(ge=0, alias="sizeBytes")


class UploadResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    project_id: str = Field(alias="projectId")
    video: VideoMetadata
    video_url: str = Field(alias="videoUrl")

File: frontend/src/types/video.ts
export interface VideoMetadata {
  filename: string
  duration: number
  width: number
  height: number
  fps: number
  videoCodec: string
  audioCodec: string | null
  hasAudio: boolean
  sizeBytes: number
}

export interface UploadResponse {
  projectId: string
  video: VideoMetadata
  videoUrl: string
}

export type ProcessingStatus =
  | 'idle'
  | 'uploading'
  | 'uploaded'
  | 'extracting_audio'
  | 'preparing_model'
  | 'transcribing'
  | 'segmenting'
  | 'ready'
  | 'error'

export interface ProjectProcessingState {
  projectId: string
  status: ProcessingStatus
  message: string
  error: string | null
}

export interface TranscriptWord {
  id: string
  text: string
  start: number
  end: number
  confidence: number | null
  emphasis: boolean
}

export interface CaptionSegment {
  id: string
  start: number
  end: number
  words: TranscriptWord[]
}

export interface CaptionDocument {
  projectId: string
  language: string
  duration: number
  words: TranscriptWord[]
  segments: CaptionSegment[]
}

export interface ProjectResponse extends UploadResponse {
  captions: CaptionDocument | null
}

export interface ApiErrorPayload {
  error?: {
    code?: string
    message?: string
  }
}
